[FEATURE][4] Add React Hook Form

mathiramilo-lit opened the pull request:
# Add React Hook Form with Zod Validation - 4

## Description

I've added React Hook Form for form handling and Zod for schema validation in the project.

## Changes Made

- Update Input component to receive and use `fieldName` and `register`
- Update Form component to use useForm with zod validation

mathiramilo-lit committed:
feat: add react hook form

- update input component to receive and use fieldName and register
- update form component to use useForm with zod validation

mathiramilo-lit committed:
feat: handle submitting and success states

mathiramilo-lit committed:
feat: add spinner variants

jsarrolt commented on components/form.tsx:
```diff
@@ -2,14 +2,90 @@
+  confirmPassword: string;
+}
+
+export const UserSchema = z
+  .object({
+    email: z
+      .string()
+      .min(1, 'Email is required.')
+      .email('Please enter a valid email address.')
+      .regex(
+        /^[a-zA-Z0-9._%+-]+@(gmail\.com|hotmail\.com|lightit\.io)$/,
+        'Email must be from one of the allowed domains: gmail.com, hotmail.com, or lightit.io.',
+      )
+      .trim(),
+    username: z
+      .string({ required_error: 'Username is required.' })
+      .min(1, 'Username is required.')
+      .toLowerCase()
+      .regex(
+        /^[a-z]+$/,
+        'Username can only contain lowercase letters, no spaces, numbers, or special characters.',
+      )
+      .min(3, 'Username must be at least 3 characters long.')
+      .max(32, 'Username must be 32 characters or less.')
+      .trim(),
+    password: z
+      .string({ required_error: 'Password is required.' })
+      .min(1, 'Password is required.')
+      .regex(
+        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]+$/,
+        'Password must contain at least one uppercase letter, one lowercase letter, and one special character.',
+      )
+      .min(8, 'Password must be at least 8 characters long.')
+      .max(64, 'Password must be 64 characters or less.'),
+    confirmPassword: z.string(),
+  })
+  .refine((data) => data.password === data.confirmPassword, {
+    path: ['confirmPassword'],
+    message: 'Passwords do not match.',
+  });
```
You can use `z.infer` for the form values:

```suggestion

export const userSchema = z
  .object({
    email: z
      .string()
      .min(1, 'Email is required.')
      .email('Please enter a valid email address.')
      .regex(
        /^[a-zA-Z0-9._%+-]+@(gmail\.com|hotmail\.com|lightit\.io)$/,
        'Email must be from one of the allowed domains: gmail.com, hotmail.com, or lightit.io.',
      )
      .trim(),
    username: z
      .string({ required_error: 'Username is required.' })
      .min(1, 'Username is required.')
      .toLowerCase()
      .regex(
        /^[a-z]+$/,
        'Username can only contain lowercase letters, no spaces, numbers, or special characters.',
      )
      .min(3, 'Username must be at least 3 characters long.')
      .max(32, 'Username must be 32 characters or less.')
      .trim(),
    password: z
      .string({ required_error: 'Password is required.' })
      .min(1, 'Password is required.')
      .regex(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]+$/,
        'Password must contain at least one uppercase letter, one lowercase letter, and one special character.',
      )
      .min(8, 'Password must be at least 8 characters long.')
      .max(64, 'Password must be 64 characters or less.'),
    confirmPassword: z.string(),
  })
  .refine((data) => data.password === data.confirmPassword, {
    path: ['confirmPassword'],
    message: 'Passwords do not match.',
  });
  
  type UserFormValuesType = z.infer<typeof userSchema>;
```

jsarrolt replied:
Also consider adding the schema in a separate file as `userSchema.ts` inside utils as a good practice 🙂

mathiramilo-lit replied:
Thanks for the advice! I’ll also move the schema to a separate file as suggested.

jsarrolt commented on components/form.tsx:
```diff
@@ -2,14 +2,90 @@
+        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]+$/,
+        'Password must contain at least one uppercase letter, one lowercase letter, and one special character.',
+      )
+      .min(8, 'Password must be at least 8 characters long.')
+      .max(64, 'Password must be 64 characters or less.'),
```
This may be a personal thing, but I prefer to check the length in the regular expression and adapt that error message to mention that the length is not correct, so if that error appears and the rest is not fulfilled either you can adjust it and not receive another error later.
But then again, is open for discussion 😄

mathiramilo-lit replied:
Good point! I hadn't thought of that. I’ll consider combining the length check into the regex. Thanks for the suggestion!

jsarrolt commented on components/form.tsx:
```diff
@@ -2,14 +2,90 @@
+  confirmPassword: string;
+}
+
+export const UserSchema = z
+  .object({
+    email: z
+      .string()
+      .min(1, 'Email is required.')
+      .email('Please enter a valid email address.')
+      .regex(
+        /^[a-zA-Z0-9._%+-]+@(gmail\.com|hotmail\.com|lightit\.io)$/,
+        'Email must be from one of the allowed domains: gmail.com, hotmail.com, or lightit.io.',
+      )
+      .trim(),
```
Maybe we can do something like this so its easier if we add more domains in the future

```suggestion
const emailDomains = ['@gmail.com', '@hotmail.com', '@lightit.io'];


export const UserSchema = z
  .object({
    email: z
      .string()
      .min(1, 'Email is required.')
      .email('Please enter a valid email address.')
      .refine((value) => {
    return emailDomains.some(domain => value.endsWith(domain));
  }, {
    message: 'The email must be valid and have a domain of @gmail.com, @hotmail.com, or @lightit.io',
  })
  .trim(),
```

jsarrolt reviewed (commented):
Great job!!

caroasti commented on app/page.tsx:
```diff
@@ -1,9 +1,9 @@
 export default function Home() {
   return (
-    <section className="flex min-h-screen w-full overflow-hidden bg-white md:max-h-[85vh] md:min-h-fit md:w-[75%] md:rounded-lg">
+    <section className="flex min-h-screen w-full overflow-hidden bg-white md:max-h-[85vh] md:w-[75%] md:rounded-lg">
```
suggestion: `md:w-[75%] `-> `md:w-3/4 `

danielaWaldeck95 commented on components/success.tsx:
```diff
@@ -0,0 +1,36 @@
+      )}
+      {...props}
+    >
+      <Image
+        src="/confetti-minimalistic.png"
+        alt="Confetti"
+        width={42}
+        height={42}
+      />
+      <p
```
use svg instead of png 
![image](https://github.com/user-attachments/assets/debab30d-ea6f-4466-ac9f-e34136a9f69b)

mathiramilo-lit committed:
fix: pr comments

- change png for svg components
- move user schema to his own folder
- infer user form values from user schema

mathiramilo-lit committed:
style: components filenames to camelCase
